### PR TITLE
Allow marking of tests as flaky #1016

### LINKED ISSUE #1016: Only retry only failing tests instead of running the complete test project again
I saw only a single failing tests in Travis CI run and the complete test project is ran again up to 3 times.

Idea to implement:
* Mark tests with an attribute and use a maven test runner for the retry policy.

### PR DESCRIPTION
**Pull request by farodin91**
Fixes #1016 

Based on stackoverflow post: https://stackoverflow.com/a/46207476/2978513

Signed-off-by: Jan Jansen <jan.jansen@gdata.de>
-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there an issue associated with this PR? Is it referenced in the commit message?
- [x] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you written and/or updated unit tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [x] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: janusgraph-test/src/main/java/org/janusgraph/testutil/RetryTemplateContext.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    @Override
+    public String getDisplayName(int invocationIndex) {
+        return "Invocation number " + invocationIndex + " (requires " + minSuccess + " success)";
+    }
+
+    @Override
+    public List<Extension> getAdditionalExtensions() {
+        List<Extension> objects = new java.util.ArrayList<>();
+        objects.add(new RetryingTestExecutionExtension(invocation, maxInvocations, minSuccess));
+        return objects;
```
Comment (porunov) [inline janusgraph-test/src/main/java/org/janusgraph/testutil/RetryTemplateContext.java]: `return Collections.singletonList(new RetryingTestExecutionExtension(invocation, maxInvocations, minSuccess))`

File: janusgraph-test/src/main/java/org/janusgraph/testutil/FlakyTest.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@TestTemplate
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.METHOD})
+@ExtendWith(FlakyTestExtension.class)
+public @interface FlakyTest {
+    int invocationCount();
```
Comment (porunov) [inline janusgraph-test/src/main/java/org/janusgraph/testutil/FlakyTest.java]: (not necessary)
Maybe add default values? I.e. `int invocationCount() default 8` and `int minSuccess() default 4`?

**Comment by farodin91** [comment]
The `testIndexReplay` fails often.

File: janusgraph-test/src/main/java/org/janusgraph/testutil/FlakyTestExtension.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import java.util.stream.Stream;
+
+public class FlakyTestExtension implements TestTemplateInvocationContextProvider {
+
+    @Override
+    public boolean supportsTestTemplate(ExtensionContext context) {
+        return AnnotationUtils.isAnnotated(context.getTestMethod(), FlakyTest.class);
+    }
+
+    public Stream<TestTemplateInvocationContext> provideTestTemplateInvocationContexts(ExtensionContext context) {
```
Comment (porunov) [inline janusgraph-test/src/main/java/org/janusgraph/testutil/FlakyTestExtension.java]: Add `@Overwrite` annotation

File: janusgraph-test/src/main/java/org/janusgraph/testutil/FlakyTestExtension.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        FlakyTest annotation = context.getTestMethod()
+            .flatMap(testMethods -> AnnotationSupport.findAnnotation(testMethods, FlakyTest.class))
+            .orElseThrow(() -> new ExtensionContextException("The extension should not be executed "
```
Comment (porunov) [inline janusgraph-test/src/main/java/org/janusgraph/testutil/FlakyTestExtension.java]: I didn't check but is it possible that we get a method here which isn't annotated with `@FlakyTest`?
I assume that JUnit checks `supportsTestTemplate` before calling `provideTestTemplateInvocationContexts`. If so, I think we can change `orElseThrow` to `get`. What do you think?

Comment (farodin91) [reply janusgraph-test/src/main/java/org/janusgraph/testutil/FlakyTestExtension.java]: This shouldn't happen. All example, i saw throw an explicit exception.

**Comment by farodin91** [comment]
@porunov I squashed everything and rebase it on the top of 0.3.

File: janusgraph-test/src/main/java/org/janusgraph/testutil/FlakyTestExtension.java (hunk 4 above)
Comment (porunov) [reply janusgraph-test/src/main/java/org/janusgraph/testutil/FlakyTestExtension.java]: I just thought that it is an unnecessary check if it isn't possible to get a method without `FlakyTest` here. If it is possible then sure we must throw `ExtensionContextException`.

Comment (farodin91) [reply janusgraph-test/src/main/java/org/janusgraph/testutil/FlakyTestExtension.java]: The extension can be used using `@ExtendWith`, If i'm right.

**Pull request merged**